From 1417382a9ffb9c8141ded377d3a4d5d94b4873b8 Mon Sep 17 00:00:00 2001
From: mengqinggang <mengqinggang@loongson.cn>
Date: Thu, 26 Dec 2024 14:14:12 +0800
Subject: [PATCH 5/5] LoongArch: Add relaxation support for call36 that jump to
 PLT entry

Part of commit a104f0a3e620 ("LoongArch: Add more relaxation support for
call36") in master.  This part is needed for 2.43 branch to fix a link
failure of QtWebEngine-6.8.1, which can be demonstrated with a reduced
test case (with -shared):

    .text
    .zero 1 << 27   # pretending we have a lot of code in .text
    .globl f        # preemptible, thus must be called via PLT
    f:
      move $a0, $r0
      ret
    g:
      pcaddu18i $t0, %call36(f)
      jr $t0

Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 bfd/elfnn-loongarch.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/bfd/elfnn-loongarch.c b/bfd/elfnn-loongarch.c
index ea480d74695..1703f807db4 100644
--- a/bfd/elfnn-loongarch.c
+++ b/bfd/elfnn-loongarch.c
@@ -5408,6 +5408,11 @@ loongarch_elf_relax_section (bfd *abfd, asection *sec,
 		    && GOT_TLS_GD_BOTH_P (tls_type))
 		symval += 2 * GOT_ENTRY_SIZE;
 	    }
+	  else if (h->plt.offset != MINUS_ONE)
+	    {
+	      sym_sec = htab->elf.splt ? htab->elf.splt : htab->elf.iplt;
+	      symval = h->plt.offset;
+	    }
 	  else if ((h->root.type == bfd_link_hash_defined
 		  || h->root.type == bfd_link_hash_defweak)
 		&& h->root.u.def.section != NULL
-- 
2.48.1

